File: rust/candid_parser/tests/assets/ok/management.rs
// This is an experimental feature to generate Rust binding from Candid.
// You may want to manually adjust some of the types.
#![allow(dead_code, unused_imports)]
use candid::{self, CandidType, Deserialize, Principal, Encode, Decode};
type Result<T> = std::result::Result<T, ic_agent::AgentError>;

#[derive(CandidType, Deserialize)]
pub enum BitcoinNetwork {
  #[serde(rename="mainnet")]
  Mainnet,
  #[serde(rename="testnet")]
  Testnet,
}
pub type BitcoinAddress = String;
#[derive(CandidType, Deserialize)]
pub struct GetBalanceRequest {
  pub network: BitcoinNetwork,
  pub address: BitcoinAddress,
  pub min_confirmations: Option<u32>,
}
pub type Satoshi = u64;
#[derive(CandidType, Deserialize)]
pub struct GetCurrentFeePercentilesRequest { pub network: BitcoinNetwork }
pub type MillisatoshiPerByte = u64;
#[derive(CandidType, Deserialize)]
pub enum GetUtxosRequestFilterInner {
  #[serde(rename="page")]
  Page(serde_bytes::ByteBuf),
  #[serde(rename="min_confirmations")]
  MinConfirmations(u32),
}
#[derive(CandidType, Deserialize)]
pub struct GetUtxosRequest {
  pub network: BitcoinNetwork,
  pub filter: Option<GetUtxosRequestFilterInner>,
  pub address: BitcoinAddress,
}
pub type BlockHash = serde_bytes::ByteBuf;
#[derive(CandidType, Deserialize)]
pub struct Outpoint { pub txid: serde_bytes::ByteBuf, pub vout: u32 }
#[derive(CandidType, Deserialize)]
pub struct Utxo { pub height: u32, pub value: Satoshi, pub outpoint: Outpoint }
#[derive(CandidType, Deserialize)]
pub struct GetUtxosResponse {
  pub next_page: Option<serde_bytes::ByteBuf>,
  pub tip_height: u32,
  pub tip_block_hash: BlockHash,
  pub utxos: Vec<Utxo>,
}
#[derive(CandidType, Deserialize)]
pub struct SendTransactionRequest {
  pub transaction: serde_bytes::ByteBuf,
  pub network: BitcoinNetwork,
}
pub type CanisterId = Principal;
#[derive(CandidType, Deserialize)]
pub struct CanisterStatusArg { pub canister_id: CanisterId }
#[derive(CandidType, Deserialize)]
pub enum CanisterStatusRetStatus {
  #[serde(rename="stopped")]
  Stopped,
  #[serde(rename="stopping")]
  Stopping,
  #[serde(rename="running")]
  Running,
}
#[derive(CandidType, Deserialize)]
pub struct DefiniteCanisterSettings {
  pub freezing_threshold: candid::Nat,
  pub controllers: Vec<Principal>,
  pub memory_allocation: candid::Nat,
  pub compute_allocation: candid::Nat,
}
#[derive(CandidType, Deserialize)]
pub struct CanisterStatusRet {
  pub status: CanisterStatusRetStatus,
  pub memory_size: candid::Nat,
  pub cycles: candid::Nat,
  pub settings: DefiniteCanisterSettings,
  pub idle_cycles_burned_per_day: candid::Nat,
  pub module_hash: Option<serde_bytes::ByteBuf>,
}
#[derive(CandidType, Deserialize)]
pub struct CanisterSettings {
  pub freezing_threshold: Option<candid::Nat>,
  pub controllers: Option<Vec<Principal>>,
  pub memory_allocation: Option<candid::Nat>,
  pub compute_allocation: Option<candid::Nat>,
}
#[derive(CandidType, Deserialize)]
pub struct CreateCanisterArg { pub settings: Option<CanisterSettings> }
#[derive(CandidType, Deserialize)]
pub struct CreateCanisterRet { pub canister_id: CanisterId }
#[derive(CandidType, Deserialize)]
pub struct DeleteCanisterArg { pub canister_id: CanisterId }
#[derive(CandidType, Deserialize)]
pub struct DepositCyclesArg { pub canister_id: CanisterId }
#[derive(CandidType, Deserialize)]
pub enum EcdsaCurve { #[serde(rename="secp256k1")] Secp256K1 }
#[derive(CandidType, Deserialize)]
pub struct EcdsaPublicKeyArgKeyId { pub name: String, pub curve: EcdsaCurve }
#[derive(CandidType, Deserialize)]
pub struct EcdsaPublicKeyArg {
  pub key_id: EcdsaPublicKeyArgKeyId,
  pub canister_id: Option<CanisterId>,
  pub derivation_path: Vec<serde_bytes::ByteBuf>,
}
#[derive(CandidType, Deserialize)]
pub struct EcdsaPublicKeyRet {
  pub public_key: serde_bytes::ByteBuf,
  pub chain_code: serde_bytes::ByteBuf,
}
#[derive(CandidType, Deserialize)]
pub enum HttpRequestArgMethod {
  #[serde(rename="get")]
  Get,
  #[serde(rename="head")]
  Head,
  #[serde(rename="post")]
  Post,
}
#[derive(CandidType, Deserialize)]
pub struct HttpHeader { pub value: String, pub name: String }
#[derive(CandidType, Deserialize)]
pub struct HttpResponse {
  pub status: candid::Nat,
  pub body: serde_bytes::ByteBuf,
  pub headers: Vec<HttpHeader>,
}
#[derive(CandidType, Deserialize)]
pub struct HttpRequestArgTransformInnerFunctionArg {
  pub context: serde_bytes::ByteBuf,
  pub response: HttpResponse,
}
candid::define_function!(pub HttpRequestArgTransformInnerFunction : (
    HttpRequestArgTransformInnerFunctionArg,
  ) -> (HttpResponse) query);
#[derive(CandidType, Deserialize)]
pub struct HttpRequestArgTransformInner {
  pub function: HttpRequestArgTransformInnerFunction,
  pub context: serde_bytes::ByteBuf,
}
#[derive(CandidType, Deserialize)]
pub struct HttpRequestArg {
  pub url: String,
  pub method: HttpRequestArgMethod,
  pub max_response_bytes: Option<u64>,
  pub body: Option<serde_bytes::ByteBuf>,
  pub transform: Option<HttpRequestArgTransformInner>,
  pub headers: Vec<HttpHeader>,
}
pub type WasmModule = serde_bytes::ByteBuf;
#[derive(CandidType, Deserialize)]
pub enum InstallCodeArgMode {
  #[serde(rename="reinstall")]
  Reinstall,
  #[serde(rename="upgrade")]
  Upgrade,
  #[serde(rename="install")]
  Install,
}
#[derive(CandidType, Deserialize)]
pub struct InstallCodeArg {
  pub arg: serde_bytes::ByteBuf,
  pub wasm_module: WasmModule,
  pub mode: InstallCodeArgMode,
  pub canister_id: CanisterId,
}
#[derive(CandidType, Deserialize)]
pub struct ProvisionalCreateCanisterWithCyclesArg {
  pub settings: Option<CanisterSettings>,
  pub specified_id: Option<CanisterId>,
  pub amount: Option<candid::Nat>,
}
#[derive(CandidType, Deserialize)]
pub struct ProvisionalCreateCanisterWithCyclesRet {
  pub canister_id: CanisterId,
}
#[derive(CandidType, Deserialize)]
pub struct ProvisionalTopUpCanisterArg {
  pub canister_id: CanisterId,
  pub amount: candid::Nat,
}
#[derive(CandidType, Deserialize)]
pub struct SignWithEcdsaArgKeyId { pub name: String, pub curve: EcdsaCurve }
#[derive(CandidType, Deserialize)]
pub struct SignWithEcdsaArg {
  pub key_id: SignWithEcdsaArgKeyId,
  pub derivation_path: Vec<serde_bytes::ByteBuf>,
  pub message_hash: serde_bytes::ByteBuf,
}
#[derive(CandidType, Deserialize)]
pub struct SignWithEcdsaRet { pub signature: serde_bytes::ByteBuf }
#[derive(CandidType, Deserialize)]
pub struct StartCanisterArg { pub canister_id: CanisterId }
#[derive(CandidType, Deserialize)]
pub struct StopCanisterArg { pub canister_id: CanisterId }
#[derive(CandidType, Deserialize)]
pub struct UninstallCodeArg { pub canister_id: CanisterId }
#[derive(CandidType, Deserialize)]
pub struct UpdateSettingsArg {
  pub canister_id: Principal,
  pub settings: CanisterSettings,
}

pub struct Service<'a>(pub Principal, pub &'a ic_agent::Agent);
impl<'a> Service<'a> {
  pub async fn bitcoin_get_balance(&self, arg0: &GetBalanceRequest) -> Result<Satoshi> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "bitcoin_get_balance").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, Satoshi)?)
  }
  pub async fn bitcoin_get_current_fee_percentiles(&self, arg0: &GetCurrentFeePercentilesRequest) -> Result<Vec<MillisatoshiPerByte>> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "bitcoin_get_current_fee_percentiles").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, Vec<MillisatoshiPerByte>)?)
  }
  pub async fn bitcoin_get_utxos(&self, arg0: &GetUtxosRequest) -> Result<GetUtxosResponse> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "bitcoin_get_utxos").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, GetUtxosResponse)?)
  }
  pub async fn bitcoin_send_transaction(&self, arg0: &SendTransactionRequest) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "bitcoin_send_transaction").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn canister_status(&self, arg0: &CanisterStatusArg) -> Result<CanisterStatusRet> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "canister_status").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, CanisterStatusRet)?)
  }
  pub async fn create_canister(&self, arg0: &CreateCanisterArg) -> Result<CreateCanisterRet> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "create_canister").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, CreateCanisterRet)?)
  }
  pub async fn delete_canister(&self, arg0: &DeleteCanisterArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "delete_canister").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn deposit_cycles(&self, arg0: &DepositCyclesArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "deposit_cycles").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn ecdsa_public_key(&self, arg0: &EcdsaPublicKeyArg) -> Result<EcdsaPublicKeyRet> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "ecdsa_public_key").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, EcdsaPublicKeyRet)?)
  }
  pub async fn http_request(&self, arg0: &HttpRequestArg) -> Result<HttpResponse> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "http_request").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, HttpResponse)?)
  }
  pub async fn install_code(&self, arg0: &InstallCodeArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "install_code").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn provisional_create_canister_with_cycles(&self, arg0: &ProvisionalCreateCanisterWithCyclesArg) -> Result<ProvisionalCreateCanisterWithCyclesRet> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "provisional_create_canister_with_cycles").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, ProvisionalCreateCanisterWithCyclesRet)?)
  }
  pub async fn provisional_top_up_canister(&self, arg0: &ProvisionalTopUpCanisterArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "provisional_top_up_canister").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn raw_rand(&self) -> Result<serde_bytes::ByteBuf> {
    let args = Encode!()?;
    let bytes = self.1.update(&self.0, "raw_rand").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, serde_bytes::ByteBuf)?)
  }
  pub async fn sign_with_ecdsa(&self, arg0: &SignWithEcdsaArg) -> Result<SignWithEcdsaRet> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "sign_with_ecdsa").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes, SignWithEcdsaRet)?)
  }
  pub async fn start_canister(&self, arg0: &StartCanisterArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "start_canister").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn stop_canister(&self, arg0: &StopCanisterArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "stop_canister").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn uninstall_code(&self, arg0: &UninstallCodeArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "uninstall_code").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
  pub async fn update_settings(&self, arg0: &UpdateSettingsArg) -> Result<()> {
    let args = Encode!(&arg0)?;
    let bytes = self.1.update(&self.0, "update_settings").with_arg(args).call_and_wait().await?;
    Ok(Decode!(&bytes)?)
  }
}
pub const CANISTER_ID : Principal = Principal::from_slice(&[]); // aaaaa-aa

File: rust/candid_parser/tests/assets/ok/management.js
import { IDL } from '@dfinity/candid';

export const bitcoin_network = IDL.Variant({
  'mainnet' : IDL.Null,
  'testnet' : IDL.Null,
});
export const bitcoin_address = IDL.Text;
export const get_balance_request = IDL.Record({
  'network' : bitcoin_network,
  'address' : bitcoin_address,
  'min_confirmations' : IDL.Opt(IDL.Nat32),
});
export const satoshi = IDL.Nat64;
export const get_current_fee_percentiles_request = IDL.Record({
  'network' : bitcoin_network,
});
export const millisatoshi_per_byte = IDL.Nat64;
export const get_utxos_request = IDL.Record({
  'network' : bitcoin_network,
  'filter' : IDL.Opt(
    IDL.Variant({ 'page' : IDL.Vec(IDL.Nat8), 'min_confirmations' : IDL.Nat32 })
  ),
  'address' : bitcoin_address,
});
export const block_hash = IDL.Vec(IDL.Nat8);
export const outpoint = IDL.Record({
  'txid' : IDL.Vec(IDL.Nat8),
  'vout' : IDL.Nat32,
});
export const utxo = IDL.Record({
  'height' : IDL.Nat32,
  'value' : satoshi,
  'outpoint' : outpoint,
});
export const get_utxos_response = IDL.Record({
  'next_page' : IDL.Opt(IDL.Vec(IDL.Nat8)),
  'tip_height' : IDL.Nat32,
  'tip_block_hash' : block_hash,
  'utxos' : IDL.Vec(utxo),
});
export const send_transaction_request = IDL.Record({
  'transaction' : IDL.Vec(IDL.Nat8),
  'network' : bitcoin_network,
});
export const canister_id = IDL.Principal;
export const definite_canister_settings = IDL.Record({
  'freezing_threshold' : IDL.Nat,
  'controllers' : IDL.Vec(IDL.Principal),
  'memory_allocation' : IDL.Nat,
  'compute_allocation' : IDL.Nat,
});
export const canister_settings = IDL.Record({
  'freezing_threshold' : IDL.Opt(IDL.Nat),
  'controllers' : IDL.Opt(IDL.Vec(IDL.Principal)),
  'memory_allocation' : IDL.Opt(IDL.Nat),
  'compute_allocation' : IDL.Opt(IDL.Nat),
});
export const ecdsa_curve = IDL.Variant({ 'secp256k1' : IDL.Null });
export const http_header = IDL.Record({
  'value' : IDL.Text,
  'name' : IDL.Text,
});
export const http_response = IDL.Record({
  'status' : IDL.Nat,
  'body' : IDL.Vec(IDL.Nat8),
  'headers' : IDL.Vec(http_header),
});
export const wasm_module = IDL.Vec(IDL.Nat8);

export const idlService = IDL.Service({
  'bitcoin_get_balance' : IDL.Func([get_balance_request], [satoshi], []),
  'bitcoin_get_current_fee_percentiles' : IDL.Func(
      [get_current_fee_percentiles_request],
      [IDL.Vec(millisatoshi_per_byte)],
      [],
    ),
  'bitcoin_get_utxos' : IDL.Func([get_utxos_request], [get_utxos_response], []),
  'bitcoin_send_transaction' : IDL.Func([send_transaction_request], [], []),
  'canister_status' : IDL.Func(
      [IDL.Record({ 'canister_id' : canister_id })],
      [
        IDL.Record({
          'status' : IDL.Variant({
            'stopped' : IDL.Null,
            'stopping' : IDL.Null,
            'running' : IDL.Null,
          }),
          'memory_size' : IDL.Nat,
          'cycles' : IDL.Nat,
          'settings' : definite_canister_settings,
          'idle_cycles_burned_per_day' : IDL.Nat,
          'module_hash' : IDL.Opt(IDL.Vec(IDL.Nat8)),
        }),
      ],
      [],
    ),
  'create_canister' : IDL.Func(
      [IDL.Record({ 'settings' : IDL.Opt(canister_settings) })],
      [IDL.Record({ 'canister_id' : canister_id })],
      [],
    ),
  'delete_canister' : IDL.Func(
      [IDL.Record({ 'canister_id' : canister_id })],
      [],
      [],
    ),
  'deposit_cycles' : IDL.Func(
      [IDL.Record({ 'canister_id' : canister_id })],
      [],
      [],
    ),
  'ecdsa_public_key' : IDL.Func(
      [
        IDL.Record({
          'key_id' : IDL.Record({ 'name' : IDL.Text, 'curve' : ecdsa_curve }),
          'canister_id' : IDL.Opt(canister_id),
          'derivation_path' : IDL.Vec(IDL.Vec(IDL.Nat8)),
        }),
      ],
      [
        IDL.Record({
          'public_key' : IDL.Vec(IDL.Nat8),
          'chain_code' : IDL.Vec(IDL.Nat8),
        }),
      ],
      [],
    ),
  'http_request' : IDL.Func(
      [
        IDL.Record({
          'url' : IDL.Text,
          'method' : IDL.Variant({
            'get' : IDL.Null,
            'head' : IDL.Null,
            'post' : IDL.Null,
          }),
          'max_response_bytes' : IDL.Opt(IDL.Nat64),
          'body' : IDL.Opt(IDL.Vec(IDL.Nat8)),
          'transform' : IDL.Opt(
            IDL.Record({
              'function' : IDL.Func(
                  [
                    IDL.Record({
                      'context' : IDL.Vec(IDL.Nat8),
                      'response' : http_response,
                    }),
                  ],
                  [http_response],
                  ['query'],
                ),
              'context' : IDL.Vec(IDL.Nat8),
            })
          ),
          'headers' : IDL.Vec(http_header),
        }),
      ],
      [http_response],
      [],
    ),
  'install_code' : IDL.Func(
      [
        IDL.Record({
          'arg' : IDL.Vec(IDL.Nat8),
          'wasm_module' : wasm_module,
          'mode' : IDL.Variant({
            'reinstall' : IDL.Null,
            'upgrade' : IDL.Null,
            'install' : IDL.Null,
          }),
          'canister_id' : canister_id,
        }),
      ],
      [],
      [],
    ),
  'provisional_create_canister_with_cycles' : IDL.Func(
      [
        IDL.Record({
          'settings' : IDL.Opt(canister_settings),
          'specified_id' : IDL.Opt(canister_id),
          'amount' : IDL.Opt(IDL.Nat),
        }),
      ],
      [IDL.Record({ 'canister_id' : canister_id })],
      [],
    ),
  'provisional_top_up_canister' : IDL.Func(
      [IDL.Record({ 'canister_id' : canister_id, 'amount' : IDL.Nat })],
      [],
      [],
    ),
  'raw_rand' : IDL.Func([], [IDL.Vec(IDL.Nat8)], []),
  'sign_with_ecdsa' : IDL.Func(
      [
        IDL.Record({
          'key_id' : IDL.Record({ 'name' : IDL.Text, 'curve' : ecdsa_curve }),
          'derivation_path' : IDL.Vec(IDL.Vec(IDL.Nat8)),
          'message_hash' : IDL.Vec(IDL.Nat8),
        }),
      ],
      [IDL.Record({ 'signature' : IDL.Vec(IDL.Nat8) })],
      [],
    ),
  'start_canister' : IDL.Func(
      [IDL.Record({ 'canister_id' : canister_id })],
      [],
      [],
    ),
  'stop_canister' : IDL.Func(
      [IDL.Record({ 'canister_id' : canister_id })],
      [],
      [],
    ),
  'uninstall_code' : IDL.Func(
      [IDL.Record({ 'canister_id' : canister_id })],
      [],
      [],
    ),
  'update_settings' : IDL.Func(
      [
        IDL.Record({
          'canister_id' : IDL.Principal,
          'settings' : canister_settings,
        }),
      ],
      [],
      [],
    ),
});

export const idlInitArgs = [];

/**
 * @deprecated Import IDL types directly from this module instead of using this factory function.
 */
export const idlFactory = ({ IDL }) => {
  const bitcoin_network = IDL.Variant({
    'mainnet' : IDL.Null,
    'testnet' : IDL.Null,
  });
  const bitcoin_address = IDL.Text;
  const get_balance_request = IDL.Record({
    'network' : bitcoin_network,
    'address' : bitcoin_address,
    'min_confirmations' : IDL.Opt(IDL.Nat32),
  });
  const satoshi = IDL.Nat64;
  const get_current_fee_percentiles_request = IDL.Record({
    'network' : bitcoin_network,
  });
  const millisatoshi_per_byte = IDL.Nat64;
  const get_utxos_request = IDL.Record({
    'network' : bitcoin_network,
    'filter' : IDL.Opt(
      IDL.Variant({
        'page' : IDL.Vec(IDL.Nat8),
        'min_confirmations' : IDL.Nat32,
      })
    ),
    'address' : bitcoin_address,
  });
  const block_hash = IDL.Vec(IDL.Nat8);
  const outpoint = IDL.Record({
    'txid' : IDL.Vec(IDL.Nat8),
    'vout' : IDL.Nat32,
  });
  const utxo = IDL.Record({
    'height' : IDL.Nat32,
    'value' : satoshi,
    'outpoint' : outpoint,
  });
  const get_utxos_response = IDL.Record({
    'next_page' : IDL.Opt(IDL.Vec(IDL.Nat8)),
    'tip_height' : IDL.Nat32,
    'tip_block_hash' : block_hash,
    'utxos' : IDL.Vec(utxo),
  });
  const send_transaction_request = IDL.Record({
    'transaction' : IDL.Vec(IDL.Nat8),
    'network' : bitcoin_network,
  });
  const canister_id = IDL.Principal;
  const definite_canister_settings = IDL.Record({
    'freezing_threshold' : IDL.Nat,
    'controllers' : IDL.Vec(IDL.Principal),
    'memory_allocation' : IDL.Nat,
    'compute_allocation' : IDL.Nat,
  });
  const canister_settings = IDL.Record({
    'freezing_threshold' : IDL.Opt(IDL.Nat),
    'controllers' : IDL.Opt(IDL.Vec(IDL.Principal)),
    'memory_allocation' : IDL.Opt(IDL.Nat),
    'compute_allocation' : IDL.Opt(IDL.Nat),
  });
  const ecdsa_curve = IDL.Variant({ 'secp256k1' : IDL.Null });
  const http_header = IDL.Record({ 'value' : IDL.Text, 'name' : IDL.Text });
  const http_response = IDL.Record({
    'status' : IDL.Nat,
    'body' : IDL.Vec(IDL.Nat8),
    'headers' : IDL.Vec(http_header),
  });
  const wasm_module = IDL.Vec(IDL.Nat8);
  return IDL.Service({
    'bitcoin_get_balance' : IDL.Func([get_balance_request], [satoshi], []),
    'bitcoin_get_current_fee_percentiles' : IDL.Func(
        [get_current_fee_percentiles_request],
        [IDL.Vec(millisatoshi_per_byte)],
        [],
      ),
    'bitcoin_get_utxos' : IDL.Func(
        [get_utxos_request],
        [get_utxos_response],
        [],
      ),
    'bitcoin_send_transaction' : IDL.Func([send_transaction_request], [], []),
    'canister_status' : IDL.Func(
        [IDL.Record({ 'canister_id' : canister_id })],
        [
          IDL.Record({
            'status' : IDL.Variant({
              'stopped' : IDL.Null,
              'stopping' : IDL.Null,
              'running' : IDL.Null,
            }),
            'memory_size' : IDL.Nat,
            'cycles' : IDL.Nat,
            'settings' : definite_canister_settings,
            'idle_cycles_burned_per_day' : IDL.Nat,
            'module_hash' : IDL.Opt(IDL.Vec(IDL.Nat8)),
          }),
        ],
        [],
      ),
    'create_canister' : IDL.Func(
        [IDL.Record({ 'settings' : IDL.Opt(canister_settings) })],
        [IDL.Record({ 'canister_id' : canister_id })],
        [],
      ),
    'delete_canister' : IDL.Func(
        [IDL.Record({ 'canister_id' : canister_id })],
        [],
        [],
      ),
    'deposit_cycles' : IDL.Func(
        [IDL.Record({ 'canister_id' : canister_id })],
        [],
        [],
      ),
    'ecdsa_public_key' : IDL.Func(
        [
          IDL.Record({
            'key_id' : IDL.Record({ 'name' : IDL.Text, 'curve' : ecdsa_curve }),
            'canister_id' : IDL.Opt(canister_id),
            'derivation_path' : IDL.Vec(IDL.Vec(IDL.Nat8)),
          }),
        ],
        [
          IDL.Record({
            'public_key' : IDL.Vec(IDL.Nat8),
            'chain_code' : IDL.Vec(IDL.Nat8),
          }),
        ],
        [],
      ),
    'http_request' : IDL.Func(
        [
          IDL.Record({
            'url' : IDL.Text,
            'method' : IDL.Variant({
              'get' : IDL.Null,
              'head' : IDL.Null,
              'post' : IDL.Null,
            }),
            'max_response_bytes' : IDL.Opt(IDL.Nat64),
            'body' : IDL.Opt(IDL.Vec(IDL.Nat8)),
            'transform' : IDL.Opt(
              IDL.Record({
                'function' : IDL.Func(
                    [
                      IDL.Record({
                        'context' : IDL.Vec(IDL.Nat8),
                        'response' : http_response,
                      }),
                    ],
                    [http_response],
                    ['query'],
                  ),
                'context' : IDL.Vec(IDL.Nat8),
              })
            ),
            'headers' : IDL.Vec(http_header),
          }),
        ],
        [http_response],
        [],
      ),
    'install_code' : IDL.Func(
        [
          IDL.Record({
            'arg' : IDL.Vec(IDL.Nat8),
            'wasm_module' : wasm_module,
            'mode' : IDL.Variant({
              'reinstall' : IDL.Null,
              'upgrade' : IDL.Null,
              'install' : IDL.Null,
            }),
            'canister_id' : canister_id,
          }),
        ],
        [],
        [],
      ),
    'provisional_create_canister_with_cycles' : IDL.Func(
        [
          IDL.Record({
            'settings' : IDL.Opt(canister_settings),
            'specified_id' : IDL.Opt(canister_id),
            'amount' : IDL.Opt(IDL.Nat),
          }),
        ],
        [IDL.Record({ 'canister_id' : canister_id })],
        [],
      ),
    'provisional_top_up_canister' : IDL.Func(
        [IDL.Record({ 'canister_id' : canister_id, 'amount' : IDL.Nat })],
        [],
        [],
      ),
    'raw_rand' : IDL.Func([], [IDL.Vec(IDL.Nat8)], []),
    'sign_with_ecdsa' : IDL.Func(
        [
          IDL.Record({
            'key_id' : IDL.Record({ 'name' : IDL.Text, 'curve' : ecdsa_curve }),
            'derivation_path' : IDL.Vec(IDL.Vec(IDL.Nat8)),
            'message_hash' : IDL.Vec(IDL.Nat8),
          }),
        ],
        [IDL.Record({ 'signature' : IDL.Vec(IDL.Nat8) })],
        [],
      ),
    'start_canister' : IDL.Func(
        [IDL.Record({ 'canister_id' : canister_id })],
        [],
        [],
      ),
    'stop_canister' : IDL.Func(
        [IDL.Record({ 'canister_id' : canister_id })],
        [],
        [],
      ),
    'uninstall_code' : IDL.Func(
        [IDL.Record({ 'canister_id' : canister_id })],
        [],
        [],
      ),
    'update_settings' : IDL.Func(
        [
          IDL.Record({
            'canister_id' : IDL.Principal,
            'settings' : canister_settings,
          }),
        ],
        [],
        [],
      ),
  });
};

/**
 * @deprecated Import IDL types directly from this module instead of using this factory function.
 */
export const init = ({ IDL }) => { return []; };
